### PR TITLE
Feat: Filter functionality on Product Page

### PR DESCRIPTION
**Pull request by abhijitdotsharma**
Following functionality added 
- Sort
- Category Select
- Price Slider
- Rating

Important Files to Check
- `src/pages/product-page`
- `src/components/filter`
- `src/components/products`
- `src/context/product-context.js`
- `src/reducer/filterReducer.js`
- `src/reducer/utilities`

Files not to check - 
- Mock Backend files were added in this branch, so they are included in this PR, can be skipped
- All the UI related code was reviewed in this PR  #2  .
But any review, would still be appreciated

### CODE REVIEW

File: src/components/filter/Filter.jsx [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        setClearSliderBtn(false)
+        filterDispatch({type: "SLIDER", payload: 5000})
+    }
+
+    return (
+        <aside class="filter-wrapper">
+            <div class="filter-container">
+                <div class="filter">
+                    <h3>Filter</h3>
+                    <span onClick={() => clearFilters()}>Clear</span>
```
Comment (munjalSaksham71) [inline src/components/filter/Filter.jsx]: ```onClick={clearFilter}```

File: src/components/filter/Filter.jsx [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                </div>
+
+                <div class="price">
+                    <h3>Price</h3>
+                    <div class="slider">
+                        {clearSliderBtn && <button onClick={() => filterDispatchAndHideClear()}>clear</button>}
```
Comment (munjalSaksham71) [inline src/components/filter/Filter.jsx]: same like above, If you are not passing any thing in arguments

File: src/components/filter/Filter.jsx [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+                                onChange={(e) => filterDispatchAndToggleClear(e)}
+                            />
+                            {sliderValue}
+                        </label>
+                    </div>
+                </div>
+
+                <div class="category">
+                    <h3>Category</h3>
+                    <label for="">
```
Comment (munjalSaksham71) [inline src/components/filter/Filter.jsx]: in React, it is htmlFor="" and please try to give some value inside it.

File: src/components/filter/Filter.jsx [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+                    {/* use RATING array later */}
+                    <h3>Rating</h3>
+                    <label class="block" for="">
+                        <input class="vertical-align-radio" type="radio" name="rating"
+                        onChange={() => filterDispatch({type: "RATING", payload: '4'})}
+                        checked={productState.rating === '4'} //what happens if I send payload: integer 4?
+                        />
+                        4 stars and above
+                    </label>
+                    {console.log("productState.rating: ", productState.rating)}
```
Comment (munjalSaksham71) [inline src/components/filter/Filter.jsx]: Remove Log

File: src/components/products/Products.jsx [hunk 5]
```diff
@@ -0,0 +1,16 @@
+import "./products.css";
+import { ProductCard} from "./ProductCard";
+
+export default function Products(props) {
```
Comment (munjalSaksham71) [inline src/components/products/Products.jsx:4]: Try to be consistent. Either use traditional Functions or arrow function

File: src/context/product-context.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+const ProductProvider = ({children}) => {
+    const [products, setProducts] = useState([]);
+
+    useEffect(()=>{
+        (async () => {
+            const res = await axios.get("/api/products/");
+            const resProducts = res.data.products;
+
+            setProducts(prev => [...prev, ...resProducts]);
```
Comment (munjalSaksham71) [inline src/context/product-context.js]: Why are u using prev here. I guess you don't need here.

File: src/components/filter/Filter.jsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+                </div>
+
+                <div class="price">
+                    <h3>Price</h3>
+                    <div class="slider">
+                        {clearSliderBtn && <button onClick={() => filterDispatchAndHideClear()}>clear</button>}
+                        <label>
+                            <input
+                                type="range"
+                                min="0"
```
Comment (munjalSaksham71) [inline src/components/filter/Filter.jsx:41]: Give some defaultValue.

**Review by munjalSaksham71 — APPROVED** [review]
Suggest some changes - Please go through them. 
One more thing in UI you can be more consistent in gaps b/w the checkbox and text.

File: src/components/filter/Filter.jsx (hunk 1 above)
Comment (abhijitdotsharma) [reply src/components/filter/Filter.jsx]: Didn't realize this, thanks for pointing out

File: src/components/products/Products.jsx (hunk 5 above)
Comment (abhijitdotsharma) [reply src/components/products/Products.jsx:4]: sure, I will be mindful next. Thanks

File: src/context/product-context.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+const ProductProvider = ({children}) => {
+    const [products, setProducts] = useState([]);
+
+    useEffect(()=>{
+        (async () => {
+            const res = await axios.get("/api/products/");
+            const resProducts = res.data.products;
+
+            setProducts(resProducts);
```
Comment (mariyasada) [inline src/context/product-context.js]: you can destructure data here like {data} then setproducts(data.products)

**Review by mariyasada — APPROVED** [review]
liked that you have written a comment on a particular line of code what it actually does.good to go

**Pull request merged**